### PR TITLE
Implement support for loading device fonts

### PR DESCRIPTION
**Pull request by evilpie**
This work is substantially based on @dinnerbone's https://github.com/Dinnerbone/ruffle/tree/fonts branch. 

However I modified this in some key areas:
- Font matching was moved completely to `load_device_font`
- Font searching was improved by supporting built-in names and having a list of fallbacks (mostly for my Linux system ;)
- The glyphs now use the proper scale
- Renamed `DirectFont` because I didn't like that name, I think `FontFace` is at least slightly better. (Suggestions welcome)

### CODE REVIEW

**Comment by MrCheeze** [comment]
Really exciting stuff. Looks like vertical spacing is not correct for fonts parsed from ttf, though. ([swf](https://mrcheeze.github.io/homestuck-ruffle-tester/flash/00253_local.swf))
image

**Comment by evilpie** [comment]
Thank you! I fixed it by changing the code for computing the `leading` value to dinnerbone's previous `face.line_gap()` instead of `ascender + descender`. I still don't fully understand that specific property.

**Comment by Dinnerbone** [comment]
One thing we're going to need for this to land is tests. We'll need a new config in the `test.toml` which lists device fonts to load (and default_font implementers) - defaulting to what we have atm. Then we can add some regression tests for font rendering :D

**Comment by MrCheeze** [comment]
Note that there's already [a dozen+ visual regression tests](https://github.com/ruffle-rs/ruffle/pull/14204/files#diff-95e4e52b5a426fd370fb75d74d86d334a681f2bf4e6d6801d4366e379d54d35d) that happen to try to render device fonts. Those should all switch from rendering with Noto Sans to rendering with the correct device font with this change, right? So that makes them bonus regression tests.

**Comment by Dinnerbone** [comment]
That's a fair point.

We could maybe make the default device font nothing (disabling device font rendering entirely unless enabled) - and then set something for each of the now failing tests. It'd save a bunch of work and make those tests less prone to changing whenever we happen to change the default.

**Comment by MrCheeze** [comment]
I always assumed it would be the opposite - no test configuration needed, just load any device fonts that the test swf asks for, same as opening it in desktop ruffle normally. Then you can just directly compare ruffle and FP's rendering, instead of having to deal with the extra complication of ruffle using Noto Sans instead of the correct font.

**Comment by Dinnerbone** [comment]
We can't do that - the tests can't assume any environmental factor, and must run everywhere. Maybe you have Times New Roman and I have a slightly different version, or an entirely different replacement font - the test should work on both our machines.

File: tests/tests/swfs/visual/fonts/device-font/test.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+num_frames = 1
+
+[fonts.regular]
+family = "Tinos"
```
Comment (evilpie) [inline tests/tests/swfs/visual/fonts/device-font/test.toml:4]: We could use any other font, maybe with an even more permissive license. I was just looking at Tinos anyway.

File: desktop/src/backends/ui.rs [hunk 2]
```diff
@@ -249,11 +256,71 @@ impl UiBackend for DesktopUiBackend {
...
-        _is_bold: bool,
-        _is_italic: bool,
-        _register: &mut dyn FnMut(FontDefinition),
+        name: &str,
+        is_bold: bool,
+        is_italic: bool,
+        register: &mut dyn FnMut(FontDefinition),
     ) {
+        let families = &match name {
+            "_serif" | "Times New Roman" | "" => vec![
```
Comment (Dinnerbone) [inline desktop/src/backends/ui.rs:265]: _serif/etc should be resolved by ruffle before it reaches the backend - opinions on what they should be should be set through `Player::set_default_font`

**Review by Dinnerbone — CHANGES_REQUESTED** [review]
We should remove the `default_font` feature from `desktop/Cargo.toml` as it no longer needs it

File: desktop/src/backends/ui.rs (hunk 2 above)
Comment (evilpie) [reply desktop/src/backends/ui.rs]: Good idea.

**Review by Dinnerbone — APPROVED** [review]
Thumbs up from me!

**Pull request merged**